tool/tsh/proxy.go

```diff
@@ -145,94 +168,141 @@ func setupJumpHost(cf *CLIConf, tc *libclient.TeleportClient, sp sshProxyParams)
 		return trace.Wrap(err)
 	}
 	return nil
 }
 
-func sshProxy(tc *libclient.TeleportClient, params sshProxyParams) error {
-	sshPath, err := getSSHPath()
+func dialSSHProxy(ctx context.Context, tc *libclient.TeleportClient, sp sshProxyParams) (net.Conn, error) {
+	remoteProxyAddr := net.JoinHostPort(sp.proxyHost, sp.proxyPort)
+
+	if !sp.tlsRouting {
+		conn, err := net.Dial("tcp", remoteProxyAddr)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return conn, nil
+	}
+
+	pool, err := tc.LocalAgent().ClientCertPool(sp.clusterName)
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
+	}
+
+	tlsConfig := &tls.Config{
+		RootCAs:            pool,
+		NextProtos:         []string{string(alpncommon.ProtocolProxySSH)},
+		InsecureSkipVerify: tc.InsecureSkipVerify,
+		ServerName:         sp.proxyHost,
 	}
-	keysDir := profile.FullProfilePath(tc.Config.KeysDir)
-	knownHostsPath := keypaths.KnownHostsPath(keysDir)
 
-	args := []string{
-		"-A",
-		// TODO: remove once we finish deprecating SHA1
-		"-o", fmt.Sprintf("PubkeyAcceptedKeyTypes=+%s", ssh.CertAlgoRSAv01),
-		"-o", fmt.Sprintf("UserKnownHostsFile=%s", knownHostsPath),
-		"-p", params.proxyPort,
-		params.proxyHost,
-		"-s",
-		fmt.Sprintf("proxy:%s:%s@%s", params.targetHost, params.targetPort, params.clusterName),
+	conn, err := tls.Dial("tcp", remoteProxyAddr, tlsConfig)
+	if err != nil {
+		return nil, trace.Wrap(err)
 	}
+	return conn, nil
+}
 
-	if tc.HostLogin != "" {
-		args = append([]string{"-l", tc.HostLogin}, args...)
+func proxySubsystemName(userHost, cluster string) string {
+	subsystem := fmt.Sprintf("proxy:%s", userHost)
+	if cluster != "" {
+		subsystem = fmt.Sprintf("%s@%s", subsystem, cluster)
 	}
+	return subsystem
+}
 
-	log.Debugf("Executing proxy command: %v %v.", sshPath, strings.Join(args, " "))
+func makeSSHClient(ctx context.Context, conn net.Conn, addr string, cfg *ssh.ClientConfig) (*tracessh.Client, error) {
+	cc, chs, reqs, err := tracessh.NewClientConn(ctx, conn, addr, cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return tracessh.NewClient(cc, chs, reqs), nil
+}
 
-	child := exec.Command(sshPath, args...)
-	child.Stdin = os.Stdin
-	child.Stdout = os.Stdout
-	child.Stderr = os.Stderr
-	return trace.Wrap(child.Run())
+func proxySession(ctx context.Context, sess *ssh.Session) error {
+	stdout, err := sess.StdoutPipe()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	stdin, err := sess.StdinPipe()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	stderr, err := sess.StderrPipe()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	errC := make(chan error)
+	go func() {
+		defer sess.Close()
+		_, err := io.Copy(os.Stdout, stdout)
+		errC <- err
+	}()
+	go func() {
+		defer sess.Close()
+		_, err := io.Copy(stdin, os.Stdin)
+		errC <- err
+	}()
+	go func() {
+		defer sess.Close()
+		_, err := io.Copy(os.Stderr, stderr)
+		errC <- err
+	}()
+	var errs []error
+	for i := 0; i < 3; i++ {
+		select {
+		case <-ctx.Done():
```

Comment:
> From what I see, when the context is canceled, then all three goroutines will still be running in the background (which is one problem), but then when one of them will return an error and try to push it into `errC` it will block, as no one is pulling from that channel (issue two).

Yeah you're right, however I'm not sure how to fix this, because the `io.Copy(stdin, os.Stdin)` goroutine never stops, even is `os.Stdin` is closed (no idea why).

For example, in the following code, closing `ctx` prints `done stderr` and `done stdout`, but hangs on the stdin goroutine.

```
var wg sync.WaitGroup
errC := make(chan error, 3)
wg.Add(1)
go func() {
	defer wg.Done()
	defer sess.Close()
	_, err := io.Copy(os.Stdout, stdout)
	errC <- err
	log.Println("done stdout")
}()
wg.Add(1)
go func() {
	defer wg.Done()
	defer os.Stdin.Close()
	_, err := io.Copy(stdin, os.Stdin)
	errC <- err
	log.Println("done stdin")
}()
wg.Add(1)
go func() {
	defer wg.Done()
	defer sess.Close()
	_, err := io.Copy(os.Stderr, stderr)
	errC <- err
	log.Println("done stderr")
}()
go func() {
	<-ctx.Done()
	os.Stdin.Close()
	sess.Close()
}()
go func() {
	wg.Wait()
	close(errC)
}()
return trace.NewAggregateFromChannel(errC, context.Background())
```

I'm not sure there is a way to do this properly where every goroutine closes before the program exits, but I'm open to any suggestions or solutions.